### PR TITLE
Fix namespacing for multiple instances of gazebo_ros2_control plugin

### PR DESCRIPTION
**Pull request by bobbleballs**
Fixes the issue described in #127 
It appears that when adding `__ns` tags to ros args on nodes that have already been namespaced, something odd happens.
I'm not too sure what happens on the rclcpp node side of things but removing the additional namespace tags on the args seems to fix the issue for me at least.
I'm running Humble on 22.04 so I've only tested against the humble branch.
I also removed the additional RCL_ROS_ARGS_FLAG in the arguments because I don't think it was doing anything.
Cheers Ben

### CODE REVIEW

**Comment by tonynajjar** [comment]
Hi @bobbleballs, usually PRs are opened targeting master and are then backported to older branches

**Comment by bobbleballs** [comment]
Hi @tonynajjar sorry, I had some issues with rebasing. I think I've sorted it now!

**Review by ahcorde — CHANGES_REQUESTED** [review]
can you provide a simple example of how to use the namespace with your changes? Maybe add some text in the README.md

**Comment by bobbleballs** [comment]
Hi @ahcorde,
I haven't changed any of the ways that you would use Namespaces before with this plugin, but now it works with multiple instances of the same plugin/robot description.
I've added to the [README.md here](https://github.com/bobbleballs/gazebo_ros2_control/tree/namespace_issue#multiple-namespaces)
Thanks!

**Comment by splatter96** [comment]
@ahcorde Could you please take another look at this?
I was stuck on this issue for multiple days until i stumbled upon the previously mentioned [thread](https://github.com/ros-controls/gazebo_ros2_control/pull/181). The patch in this MR fixed my problem with multi robot simulations in gazebo. It would be nice if this patch can be incorporated into the upcoming releases.

**Comment by DoppiaEffe94** [comment]
Hi All!
Has the mentioned fork already been merged into the master branch?
I'm having issue as well on creating multiple instances of the same robot in gazebo simulation.
Thanks.

**Comment by christophfroehlich** [comment]
@mauriz @splatter96 @DoppiaEffe94 can you please add your review to this PR (In the "Files changed" tab in the upper right corner: "Review Changes"? Compile the plugin from source, and test it with different setups.

**Review by christophfroehlich — APPROVED** [review]
The description sound reasonable to me.
I tested the existing examples (without any namespaces), maybe anyone can provide a working example with two namespaced robots?

File: doc/index.rst [hunk 1]
```diff
@@ -218,6 +218,38 @@ The following is a basic configuration of the controllers:
 .. literalinclude:: ../gazebo_ros2_control_demos/config/cart_controller.yaml
    :language: yaml
 
+
+Multiple Namespaces
+-----------------------------------------------------------
+The gazebo_ros2_control plugin can be launched in multiple namespaces. This is useful if it is desired to have multiple robots of the same description running at the same time.
```
Comment (mauriz) [inline doc/index.rst:224]: The gazebo_ros2-control plugin supports multiple namespaces

Comment (mauriz) [inline doc/index.rst:224]: Remove, the previous sentence is enough

File: doc/index.rst [hunk 2]
```diff
@@ -218,6 +218,38 @@ The following is a basic configuration of the controllers:
...
+If the namespace is to be defined in the URDF file, then you can use the ``<namespace>`` tag as detailed below:
```
Comment (mauriz) [inline doc/index.rst:225]: It is possible to directly define the namespace in the urdf file, as showed below:

File: doc/index.rst [hunk 3]
```diff
@@ -218,6 +218,38 @@ The following is a basic configuration of the controllers:
...
+      <ros>
+        <namespace>r1</namespace>
+        <remapping>/tf:=tf</remapping>
+        <remapping>/tf_static:=tf_static</remapping>
+        <remapping>diffdrive_controller/odom:=odom</remapping>
+      </ros>
+    </plugin>
+  </gazebo>
+
+Where ``r1`` is the namespace that the plugin nodes will be run in. Please note that the remapping of ``/tf`` and ``/tf_static`` is recommended only if you want to have a separate TF buffer for each namespace, e.g ``/r1/tf``
```
Comment (mauriz) [inline doc/index.rst:241]: associated to the particular instance of the plugin

File: doc/index.rst [hunk 4]
```diff
@@ -218,6 +218,38 @@ The following is a basic configuration of the controllers:
...
+
+A second method for launching in namespaces is by using the Gazebo ``spawn_entity.py`` tool and setting the ``-robot_namespace`` parameter to the script:
```
Comment (mauriz) [inline doc/index.rst:243]: A second method to launch an instance with a particular namespace is..

File: doc/index.rst [hunk 5]
```diff
@@ -218,6 +218,38 @@ The following is a basic configuration of the controllers:
...
+
+.. code-block:: bash
+
+  ros2 run gazebo_ros spawn_entity.py -entity robot_1 -robot_namespace r1 -topic robot_description
```
Comment (mauriz) [inline doc/index.rst:247]: use r1 as above

**Review by mauriz — COMMENTED** [review]
Sorry that is a quick review, just minor comments. See if that is reasonable

**Comment by christophfroehlich** [comment]
> Sorry that is a quick review, just minor comments. See if that is reasonable

Thanks for proof-reading. There is a little button in the code comment UI to include the code inside your comment, where you can propose your changes. This is then a one-click action to apply your changes for us, could you please change your comments above? 
Have you actually tested the new version, does it fix the mentioned problem?

**Comment by mauriz** [comment]
Ok sorry I am new to this review system. I will do that as well as code checking asap, no earlier than a couple of days though.. sorry

**Comment by DoppiaEffe94** [comment]
Hi @christophfroehlich !

I had the time to test the compilation and running of this PR.

Somehow the provided fork from does not compile in my ROS2 Humble version. 
For sake of precision, I just forked the package from the "humble" [branch](https://github.com/ros-controls/gazebo_ros2_control/tree/humble) and removed the mentioned [lines](https://github.com/ros-controls/gazebo_ros2_control/pull/181/commits/a18aeb6b61aaf865f2208628153dd66081796e3a), without touching the README.md for now.

With the above steps, now the package compiles without any issue and the namespacing problem seems to be solved, as you can see below:

Example for husky robot with namespace "husky_1":
gazebo_ros2_control_husky_1

and "husky_2":
gazebo_ros2_control_husky_2

You can see [here](https://github.com/DoppiaEffe94/gazebo_ros2_control/tree/namespace_issue_humble) my fork.

Do you want me to open a new PR for the merge request?

Thank you

**Comment by bobbleballs** [comment]
> Hi @christophfroehlich !
> 
> I had the time to test the compilation and running of this PR.
> 
> Somehow the provided fork from does not compile in my ROS2 Humble version. For sake of precision, I just forked the package from the "humble" [branch](https://github.com/ros-controls/gazebo_ros2_control/tree/humble) and removed the mentioned [lines](https://github.com/ros-controls/gazebo_ros2_control/pull/181/commits/a18aeb6b61aaf865f2208628153dd66081796e3a), without touching the README.md for now.
> 


Hi @DoppiaEffe94,
Could you give details on why this branch doesn't compile? It's currently merging against master so that could be why. This will be fixed when we backport.

Unfortunately I've now moved on from the position that I was at that was allowing me to test this regularly. But it was working when I left! I don't currently have a ROS install or the time to check things.

If possible we should test against master or rolling, but I understand that can be difficult.
Hopefully we can get this merged soon as it's a blocker for quite a few workflows!

**Comment by machineIntelligence** [comment]
Hi there @christophfroehlich. I am using Humble and Gazebo 11 and I am trying to get a gazebo_ros2_control example running using the Plankton-uuv_simulator repo found [here](https://github.com/davidorchansky/Plankton). I can run all of the examples in the repo, but none of them use ROS2 control. Here is a snippet of my URDF 
`    <xacro:zray_base
      namespace="$(arg namespace)"
      inertial_reference_frame="$(arg inertial_reference_frame)">
      <gazebo>
        <plugin name="$(arg namespace)_uuv_plugin" filename="libuuv_underwater_object_ros_plugin.so">
          <fluid_density>1028.0</fluid_density>
          <namespace>namespace</namespace>
          <flow_velocity_topic>hydrodynamics/current_velocity</flow_velocity_topic>
          <debug>0</debug>
          
          <xacro:zray_hydro_model namespace="$(arg namespace)"/>
        </plugin>
      
        <plugin name="gazebo_ros2_control" filename="libgazebo_ros2_control.so">
          <parameters>$(find zray_description)/config/controller_configuration.yaml</parameters>
            <robot_param>$(arg namespace)/robot_description</robot_param>
            <robot_param_node>$(arg namespace)/robot_state_publisher</robot_param_node>   
          <ros>
            <namespace>$(arg namespace)</namespace>
            <remapping>/tf:=tf</remapping>
            <remapping>/tf_static:=tf_static</remapping>
          </ros>             
        </plugin>
     </gazebo>
  </xacro:zray_base>`

After launching, if I check the node list I see /zray/robot_state_publisher, but I am getting the error 

> [gzserver-2] [ERROR] [1710189413.862622929] [zray.gazebo_ros2_control]: zray/robot_state_publisher service not available

 Additionally, checking ros2 service list I see 

> /zray/robot_state_publisher/describe_parameters
/zray/robot_state_publisher/get_parameter_types
/zray/robot_state_publisher/get_parameters
/zray/robot_state_publisher/list_parameters
/zray/robot_state_publisher/set_parameters
/zray/robot_state_publisher/set_parameters_atomically

**Comment by christophfroehlich** [comment]
@machineIntelligence you haven't written if you have compiled the code from this PR or speaking of the released version? From your gzserver output, it seems that there is a leading `/` missing: `zray/robot_state_publisher` instead of `/zray/robot_state_publisher`.

**Comment by christophfroehlich** [comment]
@DoppiaEffe94 how far did you get with it? could you provide a demo for gz_ros2_control_demos including two robot descriptions already provided with this package?

**Comment by DoppiaEffe94** [comment]
Hello @christophfroehlich.

Sorry for the delay, I was really busy and with a lot of projects going on.

I will set a demo in the next days and get back to you.

**Comment by ros-gy** [comment]
Leaving a note here, because this is a fix that is needed for my multi-robot simulation on Humble and 22.04. @ahcorde @DoppiaEffe94 Any progress lately?

File: doc/index.rst [hunk 6]
```diff
@@ -281,6 +281,38 @@ The following is a basic configuration of the controllers:
 .. literalinclude:: ../gazebo_ros2_control_demos/config/cart_controller.yaml
    :language: yaml
 
+
+Multiple Namespaces
+-----------------------------------------------------------
+The gazebo_ros2_control plugin can be launched in multiple namespaces. This is useful if it is desired to have multiple robots of the same description running at the same time.
+If the namespace is to be defined in the URDF file, then you can use the ``<namespace>`` tag as detailed below:
```
Comment (christophfroehlich) [inline doc/index.rst:288]:
```suggestion
It is possible to directly define the namespace in the URDF file, as shown below:
```

File: doc/index.rst [hunk 7]
```diff
@@ -281,6 +281,38 @@ The following is a basic configuration of the controllers:
...
+      <ros>
+        <namespace>r1</namespace>
+        <remapping>/tf:=tf</remapping>
+        <remapping>/tf_static:=tf_static</remapping>
+        <remapping>diffdrive_controller/odom:=odom</remapping>
+      </ros>
+    </plugin>
+  </gazebo>
+
+Where ``r1`` is the namespace that the plugin nodes will be run in. Please note that the remapping of ``/tf`` and ``/tf_static`` is recommended only if you want to have a separate TF buffer for each namespace, e.g ``/r1/tf``
```
Comment (christophfroehlich) [inline doc/index.rst:304]:
```suggestion
Where ``r1`` is the namespace associated to the particular instance of the plugin. Please note that the remapping of ``/tf`` and ``/tf_static`` is recommended only if you want to have a separate TF buffer for each namespace, e.g ``/r1/tf``
```

File: doc/index.rst [hunk 8]
```diff
@@ -281,6 +281,38 @@ The following is a basic configuration of the controllers:
...
+
+A second method for launching in namespaces is by using the Gazebo ``spawn_entity.py`` tool and setting the ``-robot_namespace`` parameter to the script:
```
Comment (christophfroehlich) [inline doc/index.rst:306]:
```suggestion
A second method to launch an instance with a particular namespace is by using the Gazebo ``spawn_entity.py`` tool and setting the ``-robot_namespace`` parameter to the script:
```

**Review by christophfroehlich — APPROVED** [review]
Tested it once again with the two documented approaches, and pushed my examples as well.

**Comment by christophfroehlich** [comment]
@mergifyio backport humble

**Review by ahcorde — APPROVED** [review]

**Pull request merged**